File: 1/2021.03.24/130.Surrounded Regions/python/solution.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        def DFS(i, j):
            if 0 <= i < len(board) and 0 <= j < len(board[i]) and board[i][j] == 'O':
                board[i][j] = 'A'
                DFS(i-1, j)
                DFS(i+1, j)
                DFS(i, j-1)
                DFS(i, j+1)
        
        m, n = len(board), len(board[0])
        for i in range(n):
            DFS(0, i)
            DFS(m - 1, i)
        for i in range(m):
            DFS(i, 0)
            DFS(i, n-1)
        
        for i in range(m):
            for j in range(n):
                if board[i][j] == 'A':
                    board[i][j] = 'O'
                elif board[i][j] == 'O':
                    board[i][j] = 'X'

File: 1/2021.03.24/130.Surrounded Regions/java/solution.java
class Solution {
    public void solve(char[][] board) {
        int m = board.length, n = board[0].length;
        for (int i = 0; i < m; i++) {
            DFS(board, i, 0);
            DFS(board, i, n - 1);
        }
        for (int i = 0; i < n; i++) {
            DFS(board, 0, i);
            DFS(board, m - 1, i);
        }
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] == 'A') {
                    board[i][j] = 'O';
                } else if (board[i][j] == 'O') {
                    board[i][j] = 'X';
                }
            }
        }
    }

    private void DFS(char[][] board, int i, int j) {
        if (i >= 0 && i < board.length && j >= 0 && j < board[i].length && board[i][j] == 'O') {
            board[i][j] = 'A';
            DFS(board, i - 1, j);
            DFS(board, i + 1, j);
            DFS(board, i, j - 1);
            DFS(board, i, j + 1);
        }
    }
}
